File: Python3/401 - 600/538 Convert BST to Greater Tree.py
# 92ms, 99.05%
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """

        def visit(node):
            if node:
                visit(node.right)
                node.val += self.val
                self.val = node.val
                visit(node.left)

        self.val = 0
        visit(root)
        return root
    

File: 401 - 600/538 Convert BST to Greater Tree.cc
// 39ms, 89.14%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
  private:
    int val = 0;

  public:
    TreeNode *convertBST(TreeNode *root)
    {
        visit(root);
        return root;
    }

  private:
    void visit(TreeNode *node)
    {
        if (node)
        {
            visit(node->right);
            node->val += val;
            val = node->val;
            visit(node->left);
        }
    }
};
